crates/tools/src/report_accounts.rs

```diff
@@ -0,0 +1,72 @@
+use anyhow::Result;
+use ckb_jsonrpc_types::{JsonBytes, Serialize};
+use gw_common::builtins::CKB_SUDT_ACCOUNT_ID;
+use std::path::Path;
+use tokio::task::JoinHandle;
+
+use crate::godwoken_rpc::GodwokenRpcClient;
+
+#[derive(Serialize, Debug)]
+struct Account {
+    pub id: u32,
+    pub code_hash: ckb_fixed_hash::H256,
+    pub ckb: u128,
+    pub nonce: u32,
+}
+
+type Task = JoinHandle<Result<Option<Account>, anyhow::Error>>;
+async fn producer(client: GodwokenRpcClient, tx: tokio::sync::mpsc::Sender<Task>) -> Result<()> {
+    for account_id in 0.. {
+        let client = client.clone();
+        let task = tokio::spawn(async move {
+            let script_hash = client.get_script_hash(account_id).await?;
+            if script_hash.as_bytes() == &[0u8; 32] {
+                return Ok::<_, anyhow::Error>(None);
+            }
+            let script = client
+                .get_script(script_hash.clone())
+                .await?
+                .expect("must hash script");
+            let code_hash = script.code_hash;
+            // balance
+            let addr = JsonBytes::from_vec(script_hash.as_bytes()[..20].to_vec());
+            let ckb = client.get_balance(addr, CKB_SUDT_ACCOUNT_ID).await?;
+            let nonce = client.get_nonce(account_id).await?;
+            Ok(Some(Account {
+                id: account_id,
+                code_hash,
+                ckb,
+                nonce,
+            }))
+        });
+        tx.send(task).await?;
+    }
+    Ok(())
+}
+
+pub async fn report_accounts<P: AsRef<Path>>(url: &str, output: P) -> Result<()> {
+    let client = GodwokenRpcClient::new(url);
+    let (tx, mut rx) = tokio::sync::mpsc::channel::<Task>(20);
+    // start tasks
+    let producer_handle = tokio::spawn(producer(client, tx));
+
+    // wait tasks
+    let mut total_account = 0;
+    let mut wtr = csv::Writer::from_path(output)?;
+    while let Some(task) = rx.recv().await {
+        match task.await?? {
+            Some(account) => {
+                wtr.serialize(account)?;
+                total_account += 1;
+                print!(".")
+            }
+            None => {
+                producer_handle.abort();
```

Comment:
Took me a while to understand the control flow.

There's a small caveat: the producer task is not aborted if some error happens. (It will still end when `rx` is dropped and `send` fails, and even if it does not it wouldn't really matter in a cli tool. But in long running services we should be more careful, e.g. by following the **structured concurrency** paradigm and use more reliable task joining/cancellation constructs.)

Also, could [`buffered`](https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.buffered) be used to simplify this code a bit?

Reply:
About stream buffering: https://gendignoux.com/blog/2021/04/01/rust-async-streams-futures-part1.html#ordered-buffering

Reply:
Yes, if we use `buffered` we may remove the channel, but we also use the channel to control the concurrent jobs of the 'producer', I think it is reliable enough for the cli like you said.

Reply:
The `n` parameter limits concurrency in `buffered`. No more than `n` futures will be polled concurrency and the result is produced in order. Usage is usually like this:

```rust
let stream = stream::iter(0..)
    .map(|x| async move {
        // async request with x
        x
    })
    .buffered(20)
    .take_while(...);
// Do what you want with the results in stream.
```

(Not saying that you have to use this, just providing an alternative approach.)